src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -0,0 +1,758 @@
+static constexpr std::wstring_view FALLBACK_LOCALE = L"en-us";
+
+using namespace Microsoft::Console::Render;
+
+DxFontRenderData::DxFontRenderData(::Microsoft::WRL::ComPtr<IDWriteFactory1> dwriteFactory) noexcept :
+    _dwriteFactory(dwriteFactory),
+    _glyphCell{},
+    _lineMetrics({}),
+    _boxDrawingEffect{},
+    _boxDrawingEffectItalic{}
+{
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteTextAnalyzer1> DxFontRenderData::Analyzer() noexcept
+{
+    return _dwriteTextAnalyzer;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteFontFallback> DxFontRenderData::GetSystemFontFallback()
+{
+    if (!_systemFontFallback)
+    {
+        ::Microsoft::WRL::ComPtr<IDWriteFactory2> factory2;
+        THROW_IF_FAILED(_dwriteFactory.As(&factory2));
+        factory2->GetSystemFontFallback(&_systemFontFallback);
+    }
+
+    return _systemFontFallback;
+}
+
+[[nodiscard]] til::size DxFontRenderData::GlyphCell() noexcept
+{
+    return _glyphCell;
+}
+
+[[nodiscard]] DxFontRenderData::LineMetrics DxFontRenderData::GetLineMetrics() noexcept
+{
+    return _lineMetrics;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteTextFormat> DxFontRenderData::DefaultTextFormat() noexcept
+{
+    return _dwriteTextFormat;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteFontFace1> DxFontRenderData::DefaultFontFace() noexcept
+{
+    return _dwriteFontFace;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IBoxDrawingEffect> DxFontRenderData::DefaultBoxDrawingEffect() noexcept
+{
+    return _boxDrawingEffect;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteTextFormat> DxFontRenderData::ItalicTextFormat() noexcept
+{
+    return _dwriteTextFormatItalic;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IDWriteFontFace1> DxFontRenderData::ItalicFontFace() noexcept
+{
+    return _dwriteFontFaceItalic;
+}
+
+[[nodiscard]] Microsoft::WRL::ComPtr<IBoxDrawingEffect> DxFontRenderData::ItalicBoxDrawingEffect() noexcept
+{
+    return _boxDrawingEffectItalic;
+}
+
+// Routine Description:
+// - Updates the font used for drawing
+// Arguments:
+// - desired - Information specifying the font that is requested
+// - actual - Filled with the nearest font actually chosen for drawing
+// - dpi - The DPI of the screen
+// Return Value:
+// - S_OK or relevant DirectX error
+[[nodiscard]] HRESULT DxFontRenderData::UpdateFont(const FontInfoDesired& desired, FontInfo& actual, const int dpi) noexcept
+{
+    try
+    {
+        std::wstring fontName(desired.GetFaceName());
+        DWRITE_FONT_WEIGHT weight = static_cast<DWRITE_FONT_WEIGHT>(desired.GetWeight());
+        DWRITE_FONT_STYLE style = DWRITE_FONT_STYLE_NORMAL;
+        DWRITE_FONT_STRETCH stretch = DWRITE_FONT_STRETCH_NORMAL;
+        std::wstring localeName = _GetLocaleName();
+
+        // _ResolveFontFaceWithFallback overrides the last argument with the locale name of the font,
+        // but we should use the system's locale to render the text.

```

Comment:
This isn't related to the PR itself. But I don't quite get this comment saying "`_ResolveFontFaceWithFallback` overrides the last argument with the locale name of the font". Does it?

Reply:
It does. When you call the function, you stick your requests/proposals IN.... the algorithm decides on a font for you even if it has to disrespect a few of your requests to get to something workable... and then it tells you about what it actually decided on the way OUT... so it might overwrite some of what you asked for.

And if you don't want it overridden, you need to be careful on which ref you're passing.

It looks like in this case, for some reason, I determined that we should always be telling DirectWrite to layout/draw things with the system locale, even if when we resolved a font, the font said it belonged to a different locale. I think this is because you could be like "I want a font that has zh-CN characters in it." and the system might go "uhhhh... best I can do for you is MS Gothic and it's a ja-JP sort of font" that we want to still tell the layout that the user WANTED zh-CN layout of the glyphs even though the best match font wasn't strictly a zh-CN font.

Reply:
Thanks for the explanation, Michael. Hope I can get to know more of font in DirectWrite later.